packages/tizen_rpc_port/example/client/lib/message_client.dart

```diff
@@ -0,0 +1,211 @@
+}
+
+abstract class _CallbackBase extends Parcelable {
+  _CallbackBase(this.id, this.once) {
+    sequenceId = sequenceNum++;
+  }
+
+  int id = 0;
+  bool once = false;
+  int sequenceId = 0;
+  static int sequenceNum = 0;
+
+  String get tag => '$id::$sequenceId';
+
+  Future<void> _onReceivedEvent(Parcel parcel);
+
+  @override
+  void serialize(Parcel parcel) {
+    parcel.writeInt32(id);
+    parcel.writeInt32(sequenceId);
+    parcel.writeBool(once);
+  }
+
+  @override
+  void deserialize(Parcel parcel) {
+    id = parcel.readInt32();
+    sequenceId = parcel.readInt32();
+    once = parcel.readBool();
+  }
+}
+
+abstract class NotifyCallback extends _CallbackBase {
+  NotifyCallback({bool once = false})
+      : super(_DelegateId.notifyCallback.id, once);
+
+  Future<void> onReceived(String sender, String message);
+
+  @override
+  Future<void> _onReceivedEvent(Parcel parcel) async {
+    final String sender = parcel.readString();
+    final String message = parcel.readString();
+    await onReceived(sender, message);
+  }
+}
+
+abstract class Message extends ProxyBase {
+  Message(String appid) : super(appid, 'Message');
+
+  bool _online = false;
+  final List<_CallbackBase> _delegateList = <_CallbackBase>[];
+
+  Future<void> onConnected();
+
+  Future<void> onDisconnected();
+
+  Future<void> onRejected(String errorMessage);
+
+  @override
+  @nonVirtual
+  Future<void> onConnectedEvent() async {
+    _online = true;
+    await onConnected();
+  }
+
+  @override
+  @nonVirtual
+  Future<void> onDisconnectedEvent() async {
+    _online = false;
+    await onDisconnected();
+  }
+
+  @override
+  @nonVirtual
+  Future<void> onRejectedEvent(String errorMessage) async {
+    await onRejected(errorMessage);
+  }
+
+  @override
+  @nonVirtual
+  Future<void> onReceivedEvent(Parcel parcel) async {
+    final int cmd = parcel.readInt32();
+    if (cmd != _MethodId.callback.id) {
+      return;
+    }
+
+    final int id = parcel.readInt32();
+    final int sequenceId = parcel.readInt32();
+    final bool once = parcel.readBool();
+    for (final _CallbackBase delegate in _delegateList) {
+      if (delegate.id == id && delegate.sequenceId == sequenceId) {
+        await delegate._onReceivedEvent(parcel);
+        if (delegate.once && once) {
+          _delegateList.remove(delegate);
+        }
+        break;
+      }
+    }
+  }
+
+  Parcel _consumeCommand(Port port) {
+    try {
+      final Parcel parcel = Parcel.fromPort(port);
+      final int cmd = parcel.readInt32();
+      if (cmd != _MethodId.result.id) {
+        throw Exception('Received parcel is invalid. $cmd');
+      }
+      return parcel;
+    } catch (error) {
+      return Parcel();
+    }
+  }
+
+  void disposeCallback(String tag) {
+    _delegateList.removeWhere((_CallbackBase element) => element.tag == tag);
+  }
+
+  Future<int> register(String name, NotifyCallback callback) async {
+    if (!_online) {
+      throw StateError('Must be connected first');
+    }
```

Comment:
Yes, it has to be called before invoking methods.
While calling the connect() method, the rpc-port library sends the launch request to amd(appmanager daemon) for on-demand launch. And then, it monitors the socket of the stub(server) application to connect to the server.

The implementation of the rpc-port is not equal to gRPC. The first implementation of the connect() method of the rpc-port was to send the launch request and share socket pairs.

Reply:
Thank you for your good opinion.
I think calling the connect explicitly has advantages and disadvantages.
We can know from the app developer's point of view when we are connected by explicitly calling the `connect()` and sometimes we actually want it.
I think this is the difference in design. We did not benchmark the design of `gRPC`.

Reply:
I thought `connect` could be implicitly called by each method (when not already connected). If it's by design, I'm fine with it. My little concern is that if there's a `connect` method, the user might expect a `disconnect` method as a counterpart.

Reply:
Thanks for the review!
Yes, I agree with your opinion.
So, We added `disconnect()` method on TIDL generator before.
But We didn't add in example apps because it is not used.

Reply:
Thanks for explaning why you thougth that.
We will consider improving this problem in the next version(Tizen 7.5 or 8.0).

Reply:
Thank you for the explanations!